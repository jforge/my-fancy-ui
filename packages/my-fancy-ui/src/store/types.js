// @flow

import type {StateType as CommentsStateType} from './modules/comments/types.js';

export type StateType = {
  comments: CommentsStateType
};
