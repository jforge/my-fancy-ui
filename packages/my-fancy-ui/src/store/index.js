// @flow

import type {Task as ReduxSagaInstanceType} from 'redux-saga';
import type {StateType} from './types.js';

export type StoreOptionsType = {
  initialState?: StateType | Object,
  errors?: Array<{message: string, stack: string}>,
  reduxSagaContext: 'client' | 'server' | 'universal' | 'test'
};
type StoreContextType = {
  store: any,
  rootSagaInstance: ReduxSagaInstanceType<*>
};

import {applyMiddleware, createStore as createReduxStore} from 'redux';
import createSagaMiddleware from 'redux-saga';
import {combineReduxDucks} from 'redux-lumbergh';
import serialize from 'serialize-error';
import reduxModules from './manifest.js';

export class Store {
  /**
   * Creates the redux store instance of the application.
   *
   * @param  {Object} opts The options to bootstrap the redux store.
   * @return {Object}      The object containing the finalized redux store and saga instance.
   */
  static createStore(opts: StoreOptionsType): StoreContextType {
    const {initialState, reduxSagaContext} = opts;
    const {rootReducer, rootSaga} = combineReduxDucks({
      ducks: reduxModules,
      context: reduxSagaContext
    });
    const sagaMiddleware = createSagaMiddleware();
    const rootMiddleware = applyMiddleware(sagaMiddleware);
    const store = createReduxStore(rootReducer, initialState, rootMiddleware);
    const rootSagaInstance = sagaMiddleware.run(rootSaga);

    return {store, rootSagaInstance};
  }

  /**
   * Prepare the server side props for the application.
   *
   * @return {Promise}      The Promise that resolves with the server side props of the application.
   */
  static async createServerProps(): Promise<StoreOptionsType> {
    const errors = [];

    //
    // Since this function creates the props for the client, we explicitely set the reduxSagaContext to client only,
    // to avoid e.g. duplicate HTTP requests which where already executed on the server.
    //
    let reduxSagaContext = 'client';
    let initialState = {};

    //
    // We wrap the server side state generation inside a try/catch to have the ability to fall back to an
    // server AND client saga behavior if the state generation failed.
    //
    try {
      const {store, rootSagaInstance} = Store.createStore({
        reduxSagaContext: 'server'
      });

      await rootSagaInstance.done;

      initialState = store.getState();
    } catch (e) {
      reduxSagaContext = 'universal';
      errors.push(serialize(e));
    }

    return {
      reduxSagaContext,
      errors,
      initialState
    };
  }
}
