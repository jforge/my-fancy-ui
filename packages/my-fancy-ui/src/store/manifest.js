// @flow

import * as comments from './modules/comments/';

export default [comments];
