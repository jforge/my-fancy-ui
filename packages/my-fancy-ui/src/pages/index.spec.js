// @flow

import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import logger from '@immowelt/my-fancy-ui-logger';
import {Store} from './../store/';
import IndexPage from './index.js';

describe('<IndexPage/>', () => {
  let props;

  beforeEach(() => {
    props = {
      query: {},
      errors: [],
      initialState: {},
      reduxSagaContext: 'test'
    };
  });

  it('should render correctly.', () => {
    const wrapper = shallow(<IndexPage {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should not throw errors but log them when passed from the server side.', () => {
    const spy = jest.spyOn(logger, 'error').mockImplementation(jest.fn());
    const errors = [
      {
        message: '',
        stack: ''
      }
    ];

    expect(() =>
      shallow(<IndexPage {...props} errors={errors} />)
    ).not.toThrow();
    expect(spy.mock.calls.length).toBe(1);
  });
});

describe('IndexPage.getInitialProps()', () => {
  let createServerProps;

  beforeEach(() => {
    createServerProps = jest
      .spyOn(Store, 'createServerProps')
      .mockImplementation(jest.fn());
  });

  afterEach(() => {
    // $FlowFixMe: Ignore errors since the jest type-def is out of date.
    jest.restoreAllMocks();
    jest.clearAllMocks();
  });

  it('should be a function', () => {
    expect(typeof IndexPage.getInitialProps).toBe('function');
  });

  it('should call the "createServerProps" method of the store', async () => {
    await IndexPage.getInitialProps({req: {query: {}}});

    expect(createServerProps).toHaveBeenCalledTimes(1);
  });
});
