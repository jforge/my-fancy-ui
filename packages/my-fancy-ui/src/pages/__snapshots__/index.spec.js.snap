// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexPage/> should render correctly. 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <div
    className="mdl-layout__content"
    style={
      Object {
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <div
      className="page-content"
      style={
        Object {
          "maxWidth": "640px",
        }
      }
    >
      <h2>
        Oii, mate!
      </h2>
      <Connect(CommentsList) />
    </div>
  </div>
</Provider>
`;
